File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
def solution(user_id, banned_id):
    answer = 0
    
    idx=[ [] for _ in range(len(banned_id)) ]
    for u in range(len(user_id)):
        user_id[u] = list(user_id[u])
    
    for b in range(len(banned_id)):
        banned_id[b]=list(banned_id[b])
        for i in range(len(banned_id[b])):
            if banned_id[b][i] == '*':
                idx[b].append(i)
                
    def dfs(L, tmp):
        if L == len(banned_id):
            ans.append(tmp)
            tmp = []
            return
        for u in range(len(user_id)):
            if len(user_id[u])==len(banned_id[L]):
                ban = banned_id[L][:]
                for i in idx[L]:
                    ban[i] = user_id[u][i]
                    
                if visited[u]==0 and user_id[u] == ban:
                    visited[u] = 1
                    dfs(L+1, tmp+["".join(user_id[u])])
                    visited[u] = 0
                
        
    visited=[0]*len(user_id)    
    ans = []
    dfs(0,[])
    answer = set()
    for a in ans:
        answer.add(tuple(set(sorted(a))))
    
    return len(answer)

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;
class Solution {
    static boolean[] visited;
    static int n;
    static String[] user_id;
    static String[] banned_id;
    static Set<Set<String>> set;
    public int solution(String[] user_id, String[] banned_id) {
        int answer = 0;
        this.user_id = user_id;
        this.banned_id = banned_id;
        
        List<List<Integer>> ban = new ArrayList<>();
        n = banned_id.length;
        visited = new boolean[user_id.length];
        set = new HashSet<>();
        dfs(0, new HashSet<>());
        
        return set.size();
    }
    public static void dfs(int L, Set<String> cur){
        if(L==n){
            set.add(new HashSet<>(cur));
            return;
        }
        for(int i=0;i<user_id.length;i++){
            if(visited[i]) continue;
            if(isCheck(user_id[i], banned_id[L])){
                visited[i] = true;
                cur.add(user_id[i]);
                dfs(L+1, cur);
                cur.remove(user_id[i]);
                visited[i] = false;
            }
        }
    }
    public static boolean isCheck(String user, String ban){
        if (user.length() != ban.length()) return false;
        for(int i=0; i<user.length(); i++){
            if(ban.charAt(i) != '*' && ban.charAt(i)!= user.charAt(i)){
                return false;
            }
        }
        
        
        return true;
    }
}